### PR TITLE
Some improvements to variable fixing

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Also enable fixing primitive variables during the evolution.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/VariableFixing/FixToAtmosphere.hpp [hunk 1]
```diff
@@ -52,8 +53,16 @@ class FixToAtmosphere {
     static type lower_bound() { return 0.0; }
     static constexpr OptionString help = {"Density of atmosphere"};
   };
+  /// \brief Rest mass density at which to impose the atmosphere. Should be
+  /// greater than the density of the atmosphere.
```
Comment (fmahebert) [inline src/Evolution/VariableFixing/FixToAtmosphere.hpp]: New line 23 says greater than or equal to, which I think it correct

File: src/Evolution/VariableFixing/FixToAtmosphere.hpp [hunk 2]
```diff
@@ -52,8 +53,16 @@ class FixToAtmosphere {
...
+  struct DensityCutoff {
+    using type = double;
+    static type lower_bound() { return 0.0; }
+    static constexpr OptionString help = {
+        "Density to impose atmosphere at. Must be > rho_atm"};
```
Comment (fmahebert) [inline src/Evolution/VariableFixing/FixToAtmosphere.hpp]: Again I think >= is correct

File: src/Evolution/VariableFixing/FixToAtmosphere.hpp [hunk 3]
```diff
@@ -52,8 +53,16 @@ class FixToAtmosphere {
...
+  };
 
-  using options = tmpl::list<DensityOfAtmosphere>;
+  using options = tmpl::list<DensityOfAtmosphere, DensityCutoff>;
   static constexpr OptionString help = {
```
Comment (fmahebert) [inline src/Evolution/VariableFixing/FixToAtmosphere.hpp:65]: maybe rephrase this OptionString to say "If the rest mass density is below the cutoff value, it is raised to the atmosphere value" ?

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.hpp [hunk 4]
```diff
@@ -63,6 +65,15 @@ class FixConservatives {
     static constexpr OptionString help = {
         "Minimum value of rest-mass density times lorentz factor"};
   };
+  /// \brief Cuttoff below which \f$\tilde{D}=\rho W\f$ is set to
```
Comment (fmahebert) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.hpp]: - typo: Cutoff
- I think your equation is missing a sqrt(det(g))

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.hpp [hunk 5]
```diff
@@ -63,6 +65,15 @@ class FixConservatives {
...
+  /// MinimumValueOfD
+  struct CutoffD {
+    using type = double;
+    static type lower_bound() { return 0.0; }
```
Comment (fmahebert) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.hpp]: noexcept

File: src/Evolution/VariableFixing/FixToAtmosphere.hpp [hunk 6]
```diff
@@ -52,8 +53,16 @@ class FixToAtmosphere {
     static type lower_bound() { return 0.0; }
     static constexpr OptionString help = {"Density of atmosphere"};
   };
+  /// \brief Rest mass density at which to impose the atmosphere. Should be
+  /// greater than the density of the atmosphere.
+  struct DensityCutoff {
+    using type = double;
+    static type lower_bound() { return 0.0; }
```
Comment (fmahebert) [inline src/Evolution/VariableFixing/FixToAtmosphere.hpp]: noexcept

**Review by fmahebert — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.cpp [hunk 7]
```diff
@@ -52,18 +52,30 @@ class FunctionOfLorentzFactor {
...
     : minimum_rest_mass_density_times_lorentz_factor_(
           minimum_rest_mass_density_times_lorentz_factor),
+      rest_mass_density_times_lorentz_factor_cutoff_(
+          rest_mass_density_times_lorentz_factor_cutoff),
       one_minus_safety_factor_for_magnetic_field_(
           1.0 - safety_factor_for_magnetic_field),
       one_minus_safety_factor_for_momentum_density_(
-          1.0 - safety_factor_for_momentum_density) {}
+          1.0 - safety_factor_for_momentum_density) {
+  ASSERT(
```
Comment (fmahebert) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FixConservatives.cpp]: Should this be a PARSE_ERROR?

File: src/Evolution/Conservative/UpdateConservatives.hpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+}  // namespace Parallel
+// IWYU pragma: no_forward_declare db::DataBox
+/// \endcond
+
+namespace Actions {
+/// \ingroup ActionsGroup
+/// \brief Compute the conservative variables from the primitive variables
+///
+/// Uses:
+/// - DataBox: Items in system::volume_sources::argument_tags
```
Comment (wthrowe) [inline src/Evolution/Conservative/UpdateConservatives.hpp]: `conservative_from_primative`

File: src/Evolution/VariableFixing/FixToAtmosphere.hpp [hunk 9]
```diff
@@ -36,9 +36,10 @@ namespace VariableFixing {
 /// \ingroup VariableFixingGroup
 /// \brief Fix the primitive variables to an atmosphere in low density regions
 ///
-/// If the rest mass density is below the specified value, it is
-/// raised to the specified value, and the pressure, specific internal
-/// energy (for one-dimensional equations of state), and specific
+/// If the rest mass density is below the specified value,
+/// \f$\rho_{\textrm{cutoff}}\f$ (DensityCutoff), it is raised to the
+/// \f$\rho_{\textrm{atm}}\f$ (DensityOfAtmosphere), and the pressure, specific
```
Comment (wthrowe) [inline src/Evolution/VariableFixing/FixToAtmosphere.hpp]: I'm a bit confused about what this change (splitting these parameters) is for.  It seems like in many cases this will cause the fixer to _lower_ the density to the atmosphere.

File: src/Evolution/VariableFixing/FixToAtmosphere.cpp [hunk 10]
```diff
@@ -17,13 +17,21 @@ namespace VariableFixing {
 
 template <size_t ThermodynamicDim>
 FixToAtmosphere<ThermodynamicDim>::FixToAtmosphere(
-    const double density_of_atmosphere) noexcept
-    : density_of_atmosphere_(density_of_atmosphere) {}
+    const double density_of_atmosphere, const double density_cutoff) noexcept
+    : density_of_atmosphere_(density_of_atmosphere),
+      density_cutoff_(density_cutoff) {
+  ASSERT(density_of_atmosphere_ <= density_cutoff_,
```
Comment (wthrowe) [inline src/Evolution/VariableFixing/FixToAtmosphere.cpp]: PARSE_ERROR here as well.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Evolution/VariableFixing/FixToAtmosphere.hpp (hunk 9 above)
Comment (fmahebert) [reply src/Evolution/VariableFixing/FixToAtmosphere.hpp]: Re: what this is for --- I discussed this in slack yesterday with Nils and Larry. The idea (from here https://arxiv.org/abs/1306.4953) is to make sure the atmosphere stays as an atmosphere. The previous version of the code would set smaller values to a floor, but if the floor evolved to higher density there was no longer any control of the fluid state. This version sets rho to rho_atm if rho < rho_cutoff with rho_cutoff >= rho_atm, and this means you have a buffer (between rho_atm and rho_cutoff) where the atmosphere can evolve but still be recontrolled back to atmosphere. To go from atmosphere to "real" densities, you need to jump from rho_atm to rho > rho_cutoff in one substep, but for sane choices of these parameters (i.e., small) this easily happens near resolved meaningful matter flows. I found this fixing approach helpful in my TOV experiments.

It is true that the comment is misleading, because if rho < rho_cutoff, the density could be raised _or_ lowered to rho_atm. Best to say "it is set to" I think.

File: src/Evolution/VariableFixing/FixToAtmosphere.hpp (hunk 3 above)
Comment (fmahebert) [reply src/Evolution/VariableFixing/FixToAtmosphere.hpp:65]: Will's comment below made me realize my suggestion is wrong --- "..., it is set to the atmosphere value" is correct.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp [hunk 11]
```diff
@@ -173,8 +173,9 @@ boost::optional<PrimitiveRecoveryData> NewmanHamlin::apply(
         valid_entries_in_aitken_pressure = 2;
       }
     }
-    if (fabs(current_pressure - previous_pressure) <
-        relative_tolerance_ * (current_pressure + previous_pressure)) {
+    if (current_pressure == previous_pressure or
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/NewmanHamlin.cpp]: `converged = current_pressure == previous_pressure or ...;`

**Comment by kidder** [comment]
@fmahebert @wthrowe @nilsdeppe this has been updated

**Review by nilsdeppe — DISMISSED** [review]

**Review by fmahebert — DISMISSED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
clang-tidy is complaining about Boost root finding things (we'll need to ignore)

I restarted the GCC debug builds (the usual chaos those have), and the GCC7 release because of ValenciaDivClean timeout (my guess is it starts swapping)

**Comment by nilsdeppe** [comment]
Only clang-tidy inside Boost root finders is failing. We're ignoring these failures.
